### PR TITLE
Add instrgen implementation

### PR DESCRIPTION
**Pull request by pdelewski**

### CODE REVIEW

File: instrgen/driver/main.go [hunk 1]
```diff
@@ -0,0 +1,168 @@
...
+	interfaces := alib.FindInterfaces(projectPath, packagePattern)
+	rootFunctions = append(rootFunctions, alib.FindRootFunctions(projectPath, packagePattern, "AutotelEntryPoint")...)
+	funcDecls := alib.FindFuncDecls(projectPath, packagePattern, interfaces)
+	backwardCallGraph := alib.BuildCallGraph(projectPath, packagePattern, funcDecls, interfaces)
+	fmt.Println("\n\tchild parent")
+	for k, v := range backwardCallGraph {
+		fmt.Print("\n\t", k)
+		fmt.Print(" ", v)
+	}
+	fmt.Println("")
```
Comment (dineshg13) [inline instrgen/driver/main.go:50]: Is this useful for user ?

File: instrgen/driver/main.go [hunk 2]
```diff
@@ -0,0 +1,168 @@
...
+		fmt.Println("\t" + fun.TypeHash())
+	}
+}
+
+// Parsing algorithm works as follows. It goes through all function
+// decls and infer function bodies to find call to AutotelEntryPoint
+// A parent function of this call will become root of instrumentation
+// Each function call from this place will be instrumented automatically.
+func executeCommand(command string, projectPath string, packagePattern string) error {
+	if command == "--inject" {
```
Comment (dineshg13) [inline instrgen/driver/main.go:104]: Can we use `switch` in case of `if`

File: instrgen/lib/callgraph.go [hunk 3]
```diff
@@ -0,0 +1,392 @@
...
+
+func getPkgs(projectPath string, packagePattern string, fset *token.FileSet) ([]*packages.Package, error) {
+	cfg := &packages.Config{Fset: fset, Mode: LoadMode, Dir: projectPath}
+	pkgs, err := packages.Load(cfg, packagePattern)
+	var packageSet []*packages.Package
+	if err != nil {
+		return nil, err
+	}
+	for _, pkg := range pkgs {
+		fmt.Println("\t", pkg)
```
Comment (dineshg13) [inline instrgen/lib/callgraph.go:56]: Can u add more meaningful println , like `finding packages` etc ?

**Review by dineshg13 — COMMENTED** [review]
Can you please add how to run this instrument ? And please trim printlns. 

When i run in macos or linux am getting this error ( go version go1.19.5 darwin/arm64 ) 

```
./instrgen --inject ./testdata/basic/fib.go ./                                                                                                                                                                                                                                                                                                                                                                                                                                   ─╯
autotel compiler
FindInterfaces
FindFuncDecls
BuildCallGraph

        child parent

2023/02/05 07:29:29 err: fork/exec /opt/homebrew/bin/go: not a directory: stderr:
```

**Comment by pdelewski** [comment]
> Can you please add how to run this instrument ? And please trim printlns.
> 
> When i run in macos or linux am getting this error ( go version go1.19.5 darwin/arm64 )
> 
> ```
>` ./instrgen --inject ./testdata/basic/fib.go ./  `                                                                                                                                                                                                                                                                                                                                                                                                                                 ─╯
> autotel compiler
> FindInterfaces
> FindFuncDecls
> BuildCallGraph
> 
>         child parent
> 
> 2023/02/05 07:29:29 err: fork/exec /opt/homebrew/bin/go: not a directory: stderr:
> ```

@dineshg13 works on packages so second parameter must be path not file, so in this case ` ./instrgen --inject ./testdata/basic/ ./  ` will work, however it will inject into all files in this directory. I grouped few simple programs under one sub-directory due to test execution speedup

File: instrgen/lib/callgraph.go (hunk 3 above)
Comment (pdelewski) [reply instrgen/lib/callgraph.go:56]: Yes

File: instrgen/driver/main.go (hunk 2 above)
Comment (pdelewski) [reply instrgen/driver/main.go:104]: Yes

File: instrgen/driver/main.go (hunk 1 above)
Comment (pdelewski) [reply instrgen/driver/main.go:50]: This is useful for debugging purposes. I left just for now, but of course we can get rid of this printfs.

File: instrgen/rtlib/rtlib.go [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+
+// NewResource returns a resource describing this application.
+func NewResource() *resource.Resource {
+	r, _ := resource.Merge(
+		resource.Default(),
+		resource.NewWithAttributes(
+			semconv.SchemaURL,
+			semconv.ServiceNameKey.String("fib"),
+			semconv.ServiceVersionKey.String("v0.1.0"),
+			attribute.String("environment", "demo"),
```
[the remark this answers is not in the document]
Comment (pdelewski) [reply instrgen/rtlib/rtlib.go:82]: Thx. That's kind of leftover. I have to think what to do with it.

**Comment by MrAlias** [comment]
SIG meeting: plan is to add a readme to this.

**Review by MrAlias — APPROVED** [review]
This looks good enough to start with. I've found a fair amount of issues, but I don't think any of them fundamentally block this from being merged as a starting point.

Thanks for getting this all together @pdelewski! :tada:

File: CHANGELOG.md [hunk 5]
```diff
@@ -72,6 +72,7 @@ The next release will require at least [Go 1.19].
 - Add missing recommended AWS Lambda resource attributes `faas.instance` and `faas.max_memory` in `go.opentelemetry.io/contrib/detectors/aws/lambda`. (#3148)
 - Improve documentation for `go.opentelemetry.io/contrib/samplers/jaegerremote` by providing examples of sampling endpoints. (#3147)
 - Add `WithServerName` to `go.opentelemetry.io/contrib/instrumentation/net/http/otelhttp` to set the primary server name of a `Handler`. (#3182)
+- Add `instrgen` implementation. (#3108)
```
[the remark this answers is not in the document]
Comment (pdelewski) [reply CHANGELOG.md]: done

**Review by dineshg13 — APPROVED** [review]
This is a good start. I was able to instrument a sample HTTP server using this. @pdelewski Thanks for the contribution.

**Comment by pdelewski** [comment]
@MrAlias Seems that we can merge it.

**Comment by MrAlias** [comment]
Merging as this has 2 reviews from the @open-telemetry/go-instrumentation-approvers group.

**Pull request merged**